multiobj_train.py

```diff
@@ -0,0 +1,301 @@
+    aux_itr = iterators.GroupedIterator(
+        aux_itr, update_freq, bottomless=True)
+
+    extra_meters = collections.defaultdict(lambda: AverageMeter())
+    valid_subsets = args.valid_subset.split(',')
+    max_update = args.max_update or math.inf
+    for i, samples in enumerate(progress, start=epoch_itr.iterations_in_epoch):
+        # Record gradients from auxiliary data
+        aux_samples = next(aux_itr)
+        trainer.train_step(aux_samples, update_params=False)
+        # if hasattr(trainer.optimizer, "save_auxiliary"):
+        trainer.optimizer.save_auxiliary()
+
+        log_output = trainer.train_step(samples)
+        if log_output is None:
+            continue
+
+        # log mid-epoch stats
+        stats = get_training_stats(trainer)
+        for k, v in log_output.items():
+            if k in ['loss', 'nll_loss', 'ntokens', 'nsentences', 'sample_size']:
+                continue  # these are already logged above
+            if 'loss' in k:
+                extra_meters[k].update(v, log_output['sample_size'])
+            else:
+                extra_meters[k].update(v)
+            stats[k] = extra_meters[k].avg
+        progress.log(stats, tag='train', step=stats['num_updates'])
+
+        # ignore the first mini-batch in words-per-second calculation
+        if i == 0:
+            trainer.get_meter('wps').reset()
+
+        num_updates = trainer.get_num_updates()
+        if (
+            not args.disable_validation
+            and args.save_interval_updates > 0
+            and num_updates % args.save_interval_updates == 0
+            and num_updates > 0
+        ):
+            valid_losses = validate(
+                args, trainer, task, epoch_itr, valid_subsets)
+            checkpoint_utils.save_checkpoint(
+                args, trainer, epoch_itr, valid_losses[0])
+
+        if num_updates >= max_update:
+            break
+
+    # log end-of-epoch stats
+    stats = get_training_stats(trainer)
+    for k, meter in extra_meters.items():
+        stats[k] = meter.avg
+    progress.print(stats, tag='train', step=stats['num_updates'])
+
+    # reset training meters
+    for k in [
+        'train_loss', 'train_nll_loss', 'wps', 'ups', 'wpb', 'bsz', 'gnorm', 'clip',
+    ]:
+        meter = trainer.get_meter(k)
+        if meter is not None:
+            meter.reset()
+
+
+def train(args, trainer, task, epoch_itr, epoch_aux_itr):
+    """Train the model for one epoch."""
+
+    # Update parameters every N batches
+    if epoch_itr.epoch <= len(args.update_freq):
+        update_freq = args.update_freq[epoch_itr.epoch - 1]
+    else:
+        update_freq = args.update_freq[-1]
+
+    # Initialize data iterator
+    itr = epoch_itr.next_epoch_itr(
+        fix_batches_to_gpus=args.fix_batches_to_gpus)
+    itr = iterators.GroupedIterator(itr, update_freq)
+    progress = progress_bar.build_progress_bar(
+        args, itr, epoch_itr.epoch, no_progress_bar='simple',
+    )
+
+    # Auxiliary iterator
```

Comment:
I wonder whether we need a separate iterator, vs. a single iterator on a multi-dataset, e.g. https://github.com/pytorch/fairseq/blob/master/fairseq/data/round_robin_zip_datasets.py.

Reply:
Good point. To be honest I used 2 iterators because I didn't want to write another Dataset class. I don't think RoundRobin fits the bill completely since it matches the length of the longest dataset. Not sure wheter one solution is better than the other tbh

Reply:
Actually the current version probably gives greater control over the batch size of each data source so it's probably better